Add pattern to merge/simplify split-cat

devashishshankar opened the pull request:
Summary:
In simple cases, both split and cat node can be removed in a "split->cat" pattern. However, there are various cases where they can't simply be removed and we need to simplify split/ add transforms before cat. Some such cases are:
* Split-dim != cat-dim (but equal split)
* Final node: cat vs stack
* Final node has additional args
* Shuffling of args between split/cat
* Some final nodes are non-(cat/stack)

For more details, please refer to https://docs.google.com/presentation/d/1SxBuY_FZfljSlX6i8slRNgP2CsUCICP0o4qe8cNNX8U/edit#slide=id.g232e9a90f64_0_273 (slides 8-15)

Differential Revision: D45452404



cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -189,9 +192,385 @@ def merge_splits(
+        self.erase_old_nodes(graph, split_node, next_users)
+
+    def find_next_users(self, split_node):
+        next_users = set()
```
Sets are not determinstic in Python.

devashishshankar replied:
The order of `next_users` doesn't matter. `next_users` are essentially things like `cat`, `stack` nodes (not getitems), so we can replace them in any order

jansel replied:
Replacing them in random order will make the output code of `torch.compile()` non-determinstic.  We consider that a bug.

jansel reviewed (changes requested):
This is done in a pretty ad hoc way.

1) Can we break it up into smaller steps?
2) Can we use more of the pattern matcher infra?  I worry there will be subtle issues hiding in places.

devashishshankar commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -215,3 +594,16 @@ def _split_cat_init():
         )
         pattern.register(normalize_split_pass)
+
+    for pattern in [
+        CallFunctionVarArgs(
+            torch.cat,
+        ),
+        CallFunctionVarArgs(
+            torch.stack,
+        ),
+    ]:
+        pattern = MergeSplitCat(
+            pattern=pattern, extra_check=config_flag("split_cat_fx_passes")
+        )
+        pattern.register(merge_split_cat_pass)
```
That's a simple change that I can make. Although I wanted a class structure, since there are so many functions - class helps group all related functions together. And I'm not sure if the decorator would work on a class method.

devashishshankar committed:
Add pattern to merge/simplify split-cat (#100713)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/100713

In simple cases, both split and cat node can be removed in a "split->cat" pattern. However, there are various cases where they can't simply be removed and we need to simplify split/ add transforms before cat. Some such cases are:
* Split-dim != cat-dim (but equal split)
* Final node: cat vs stack
* Final node has additional args
* Shuffling of args between split/cat
* Some final nodes are non-(cat/stack)

For more details, please refer to https://docs.google.com/presentation/d/1SxBuY_FZfljSlX6i8slRNgP2CsUCICP0o4qe8cNNX8U/edit#slide=id.g232e9a90f64_0_273 (slides 8-15)

Test Plan:
# Internal:

New build command:
```
buck2 kill && NO_BUCKD=1 buck2 run mode/opt //caffe2/torch/fb/training_toolkit/scripts:pyper_canary -- -f 436510913   --ads_dper3=force --training_platform=force  --modelstore.model_publish_workflows=modelstore.model_publish_workflows:pyper-prod  --dper_utilities=dper_utilities:pyper-prod --model_processing=model_processing:pyper-prod  2>&1 | tee /tmp/build.txt
```

Flow runs:

## 1. Baseline PT2 LWB run (prod pkg):

https://www.internalfb.com/mlhub/pipelines/runs/fblearner/436510829

## 2. PT2 LWB (current diff's pkg)

https://www.internalfb.com/mlhub/pipelines/runs/fblearner/436510913

## 3. PT2 LWB split_cat_fx_passes (current diff's pkg)

* enable split_cat_fx_passes
* disable `enable_split_cat_optimization` (old fxt)

https://www.internalfb.com/mlhub/pipelines/runs/fblearner/436788866

Comparison result: https://www.internalfb.com/mlhub/flowruns/compare?baseline_run=436510913&compare_to[0]=436788866&all_runs[0]=436510913&all_runs[1]=436788866&src=other

Blue = with current change (PT2-split cat)
Orange = with split_cat_fxt

Note: QPS numbers are lower as this is Light weight baseline

{F988250031}

Differential Revision: D45452404

fbshipit-source-id: 9970d0b43a490dd79c933646bccf4b5a409a07c5
